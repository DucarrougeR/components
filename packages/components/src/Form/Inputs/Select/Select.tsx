import React, { forwardRef, Ref } from 'react'
// import styled, { StyledComponent } from 'styled-components'
// import { CustomizableAttributes, palette } from '@looker/design-tokens'
// import { Box, BoxProps } from '../../../Layout/Box'
// import { InputProps } from '../InputProps'
import styled from 'styled-components'
import {
  border,
  typography,
  layout,
  CompatibleHTMLProps,
  CustomizableAttributes,
  space,
  SpaceProps,
} from '@looker/design-tokens'
import { ValidationType } from '../../ValidationMessage'

const renderOptions = (options: OptionsType<SelectOptionProps>) => {
  return options.map(option => (
    <option key={option.value} value={option.value}>
      {option.label}
    </option>
  ))
}

const renderOptGroups = (
  optionGroups: GroupedOptionsType<SelectOptionProps>
) => {
  return optionGroups.map(option => {
    return (
      <optgroup key={option.key} label={option.key}>
        {renderOptions(option.options)}
      </optgroup>
    )
  })
}

export const CustomizableSelectAttributes: CustomizableAttributes = {
  borderRadius: 'medium',
  fontSize: 'small',
  height: '28px',
  px: 'xsmall',
  py: 'none',
}

export type OptionsType<OptionType> = OptionType[]

export interface GroupType<OptionType> {
  options: OptionsType<OptionType>
  [key: string]: any
}

export type GroupedOptionsType<UnionOptionType> = Array<
  GroupType<UnionOptionType>
>

export interface SelectOptionProps {
  label: string
  value: string
}

export interface SelectProps
  extends SpaceProps,
    CompatibleHTMLProps<HTMLSelectElement> {
  options?:
    | OptionsType<SelectOptionProps>
    | GroupedOptionsType<SelectOptionProps>
  /**
   * Include a blank item as the first entry. If placeholder is specified it will be the label.
   * @default true
   */
  includeBlank?: boolean
  /**
   * Displays an example value or short hint to the user. Should not replace a label.
   */
  placeholder?: string
  validationType?: ValidationType
}

const SelectComponent = forwardRef(
  (
    {
      includeBlank = true,
      options,
      placeholder,
      validationType,
      defaultValue: propsDefault,
      value,
      ...props
    }: SelectProps,
    ref: Ref<HTMLSelectElement>
  ) => {
    // Gracefully deal with situation where `value` prop is set but `onChange` is not.
    const defaultValue =
      propsDefault || (value && !props.onChange) ? value : undefined

    const optionElements =
      !options || options.length === 0
        ? null
        : Object.prototype.hasOwnProperty.call(options[0], 'key')
        ? renderOptGroups(options as GroupedOptionsType<SelectOptionProps>)
        : renderOptions(options as OptionsType<SelectOptionProps>)

    return (
      <SelectBase
        defaultValue={defaultValue ? defaultValue.toString() : undefined}
        value={defaultValue ? undefined : value}
        {...props}
        ref={ref}
      >
        {includeBlank && <option value="">{placeholder}</option>}
        {optionElements}
      </SelectBase>
    )
  }
)

//
// @TODO - Should be properly imported from `Caret Down.svg`
// import caretDownIcon from '../../../../icons/svg/Caret Down.svg'
//
const indicatorRaw = `<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M7.41 8L12 12.58L16.59 8L18 9.41L12 15.41L6 9.41L7.41 8Z" fill="#1C2125"/>
</svg>`
const indicatorSize = '1rem'
const indicatorPadding = '.25rem'
const indicator = (color: string) =>
  window.btoa(indicatorRaw.replace('#1C2125', color))

// NOTE: Styling Selects is very complex
//  See reference artice for background: https://www.filamentgroup.com/lab/select-css.html
//  This component will likely be replaced with a React Select powered version

const SelectBase = styled.select.attrs((props: SelectProps) => ({
  borderRadius: CustomizableSelectAttributes.borderRadius,
  fontSize: CustomizableSelectAttributes.fontSize,
  height: props.py || props.p ? undefined : CustomizableSelectAttributes.height,
  px: props.p || CustomizableSelectAttributes.px,
  py: props.p || CustomizableSelectAttributes.py,
  type: 'text',
  ...props,
}))<SelectProps>`
  background: ${props =>
    props.validationType === 'error'
      ? props.theme.colors.palette.red000
      : props.theme.colors.palette.white};
  border: solid 1px ${props => props.theme.colors.palette.charcoal300};

  appearance: none;

  background-image:
    url(data:image/svg+xml;base64,
    ${props => indicator(props.theme.colors.palette.charcoal500)}),
    linear-gradient(to bottom, ${props =>
      props.theme.colors.palette.white} 0%, ${props =>
  props.theme.colors.palette.white} 100%);

  background-repeat: no-repeat, repeat;
  background-position: right ${indicatorPadding} center, 0 0;
  background-size: ${indicatorSize}, 100%;

  padding-right: calc(2 * ${indicatorPadding} + ${indicatorSize});

  &::-ms-expand {
    display: none;
  }
  ${border}
  ${layout}
  ${typography}
  ${space}
`

export const Select = styled(SelectComponent)``
