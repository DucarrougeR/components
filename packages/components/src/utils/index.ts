export * from './GlobalStyle'
export * from './useToggle'
