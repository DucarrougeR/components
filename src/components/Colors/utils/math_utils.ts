/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
