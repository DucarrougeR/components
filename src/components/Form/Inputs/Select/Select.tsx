import * as React from 'react'
import { Theme } from '../../../../style'
import { CustomizableAttributes } from '../../../../types/attributes'
import { Box, BoxProps } from '../../../Box'
import { InputProps } from '../InputProps'

export type OptionsType<OptionType> = OptionType[]

export interface GroupType<OptionType> {
  options: OptionsType<OptionType>
  [key: string]: any
}

export type GroupedOptionsType<UnionOptionType> = Array<
  GroupType<UnionOptionType>
>

export interface SelectOptionProps {
  label: string
  value: string
}

export interface SelectProps extends BoxProps<HTMLSelectElement>, InputProps {
  /**
   * Specifies value of the input field.
   */
  value?: string
  /*
   * Specifies the options that should be available to select
   */
  options?:
    | OptionsType<SelectOptionProps>
    | GroupedOptionsType<SelectOptionProps>
  /**
   * Include a blank item as the first entry. If placeholder is specified it will be the label.
   * @default true
   */
  includeBlank?: boolean
  /**
   * Displays an example value or short hint to the user. Should not replace a label.
   */
  placeholder?: string
  theme?: Theme
}

export const Select: React.SFC<SelectProps> = ({
  includeBlank = true,
  options,
  placeholder,
  validationType,
  ...props
}) => {
  const handleValidationType = () => {
    switch (validationType) {
      case 'error':
        return 'palette.red000'
      default:
        return undefined
    }
  }

  // Gracefully deal with situation where `value` prop is set but `onChange` is not.
  const defaultValue =
    props.defaultValue || (props.value && !props.onChange)
      ? props.value
      : undefined

  const optionElements = !options
    ? null
    : options[0].hasOwnProperty('key')
      ? renderOptGroups(options as GroupedOptionsType<SelectOptionProps>)
      : renderOptions(options as OptionsType<SelectOptionProps>)

  return (
    <Box
      is="select"
      bg={handleValidationType()}
      border="solid 1px"
      borderColor="palette.charcoal300"
      borderRadius={CustomizableSelectAttributes.borderRadius}
      fontSize={CustomizableSelectAttributes.fontSize}
      height={CustomizableSelectAttributes.height}
      px={CustomizableSelectAttributes.px}
      py={CustomizableSelectAttributes.py}
      type="text"
      {...props}
      defaultValue={defaultValue}
      value={defaultValue ? undefined : props.value}
    >
      {includeBlank && <option>{placeholder}</option>}
      {optionElements}
    </Box>
  )
}

const renderOptions = (options: OptionsType<SelectOptionProps>) => {
  return options.map(option => (
    <option key={option.value} value={option.value}>
      {option.label}
    </option>
  ))
}

const renderOptGroups = (
  optionGroups: GroupedOptionsType<SelectOptionProps>
) => {
  return optionGroups.map(option => {
    return (
      <optgroup key={option.key} label={option.key}>
        {renderOptions(option.options)}
      </optgroup>
    )
  })
}

export const CustomizableSelectAttributes: CustomizableAttributes = {
  borderRadius: 'medium',
  fontSize: 'small',
  height: '28px',
  px: 'xsmall',
  py: 'none',
}
