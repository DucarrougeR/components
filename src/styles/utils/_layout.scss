// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
