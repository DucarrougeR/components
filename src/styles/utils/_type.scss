// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
