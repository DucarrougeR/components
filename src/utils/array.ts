/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
