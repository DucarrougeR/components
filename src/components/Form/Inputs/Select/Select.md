```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Inputs/Select.tsx"
  feedbackTitle="Select Component Feedback"
/>
```

### Default Select

The `<Select />` component renders a single select element on the page, with no accompanying label. It is generally used to construct higher-order components like the `<FieldText />`. If you are building a form, you probably want to use `<FieldSelect />` instead as it provides label and validation support.

```js
import { Select } from './Select'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<>
  <Select options={options} />
  <Select options={options} value="gouda" ml="large" />
  <Select placeholder="awesome!" />
  <Select
    value="Some Value"
    options={[
      { value: '1', label: 'thing' },
      { value: 'Some Value', label: "Some Value's Label" },
      { value: '2', label: 'other' },
    ]}
  />
</>
```

#### Option Groups

```js
import { Select } from './Select'
const options = [
  {
    key: 'Italian',
    options: [
      { value: 'asiago', label: 'Asiago' },
      { value: 'ricotta', label: 'Ricotta' },
      { value: 'grana', label: 'Grana' },
    ],
  },
  {
    key: 'American',
    options: [
      { value: 'american', label: 'American (not actually cheese)' },
      { value: 'colby-jack', label: 'Colby-Jack' },
      { value: 'swiss', label: 'Swiss' },
      { value: 'vermont', label: 'Vermont' },
    ],
  },
]
;<Select options={options} value="american" />
```

### Name and ID

A name and ID can be specified in the `<Select />` component. Names are important if the input is used in the context of a form, in which case a name is required for the value of the input to be captured.

```js
import { Select } from './Select'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<Select name="someName" id="someId" options={options} />
```

### Disabled Property

Use the disable property to make an input field uneditable.

```js
import { Select } from './Select'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<>
  <Select disabled options={options} />
  <Select value="gouda" disabled options={options} ml="large" />
</>
```

### Placeholders & `includeBlank`

Placeholders are used to give a hint to the user of the expected value for the input. They should not be used as a complete replacement of labels. The `includeBlank` property (default: true) includes an inital option without a value.

```js
import { Select } from './Select'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<>
  <Select placeholder="Select your cheese of choice..." options={options} />
  <Select includeBlank={false} options={options} ml="large" />
</>
```

### ReadOnly Property

As the name suggests, "readOnly" makes the text uneditable.

```js
import { Select } from './Select'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<Select value="gouda" readOnly options={options} />
```

### Required Property

When set for `<Select />` in the context of a form, the input field must be filled out before submitting the form.

```js
import { Select } from './Select'
import { Button } from '../../../Button'
const options = [
  { value: 'cheddar', label: 'Cheddar' },
  { value: 'gouda', label: 'Gouda' },
  { value: 'swiss', label: 'Swiss' },
]
;<form>
  <Select required options={options} />
  <br />
  <br />
  <Button>Submit</Button>
</form>
```
