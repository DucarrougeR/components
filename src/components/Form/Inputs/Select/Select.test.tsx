import 'jest-styled-components'
import * as React from 'react'
import { mountWithTheme } from '../../../../../test/utils/create_with_theme'
import { assertSnapshot } from '../../../../../test/utils/snapshot'
import { Select } from './Select'

test('Select default', () => {
  assertSnapshot(<Select />)
})

test('Select with name and id', () => {
  assertSnapshot(<Select name="Bob" id="Bobby" />)
})

test('Select should accept disabled', () => {
  assertSnapshot(<Select disabled />)
})

test('Select with a placeholder', () => {
  assertSnapshot(<Select placeholder="I am a placeholder" />)
})

test('Select should accept readOnly', () => {
  assertSnapshot(<Select readOnly />)
})

test('Select should accept required', () => {
  assertSnapshot(<Select required />)
})

test('Select with a value', () => {
  const options = [{ value: 'Some Value', label: "Some Value's Label" }]
  assertSnapshot(<Select value="Some value" options={options} />)
})

test('Select with aria-describedby', () => {
  assertSnapshot(<Select aria-describedby="some-id" />)
})

test('Select with an error validation', () => {
  assertSnapshot(<Select validationType="error" />)
})

test('Should trigger onChange handler', () => {
  let counter = 0
  const handleChange = () => counter++

  const wrapper = mountWithTheme(<Select onChange={handleChange} />)

  wrapper.find('select').simulate('change', { target: { value: '' } })
  expect(counter).toEqual(1)
})
