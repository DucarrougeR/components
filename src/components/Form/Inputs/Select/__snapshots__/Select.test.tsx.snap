// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select default 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  type="text"
>
  <option />
</select>
`;

exports[`Select should accept disabled 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  disabled={true}
  type="text"
>
  <option />
</select>
`;

exports[`Select should accept readOnly 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  readOnly={true}
  type="text"
>
  <option />
</select>
`;

exports[`Select should accept required 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  required={true}
  type="text"
>
  <option />
</select>
`;

exports[`Select with a placeholder 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  type="text"
>
  <option>
    I am a placeholder
  </option>
</select>
`;

exports[`Select with a value 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  defaultValue="Some value"
  type="text"
>
  <option />
  <option
    value="Some Value"
  >
    Some Value's Label
  </option>
</select>
`;

exports[`Select with an error validation 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  background-color: #FFF2F4;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  type="text"
>
  <option />
</select>
`;

exports[`Select with aria-describedby 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  aria-describedby="some-id"
  className="c0"
  type="text"
>
  <option />
</select>
`;

exports[`Select with name and id 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<select
  className="c0"
  id="Bobby"
  name="Bob"
  type="text"
>
  <option />
</select>
`;
